File: Month-2/Week-6/day-40/c++/main.cpp
/**
 * DAY-40 C++ version
 * 
 * Just run 'make' to compile and run all the tests
 */
#define DOCTEST_CONFIG_IMPLEMENT_WITH_MAIN
#include "../../../../dependencies/c++/doctest.h" //https://github.com/onqtam/doctest

/**
 * Check if is possible to sort array using the following operation any number of times: 
 * Choose any 3 consecutive indices and rotate their elements in such a way that ABC ➞ BCA ➞ CAB ➞ ABC.
 * 
 * @param a Sequence of natural numbers incrementing from 1
 * @param true if it is possible
 */ 
bool larryArray(std::vector<int> a) {
    int inversions = 0;
    for (int i = 0; i < a.size(); i++) {
        for (int j = i + 1; j < a.size(); j++) {
            inversions += a[i] > a[j] ? 1 : 0;
        }
    }
    return inversions % 2 == 0;
}

/**
 * Tests
 */
TEST_CASE("Tests")
{
    CHECK(larryArray({1, 6, 5, 2, 4, 3}) == true);
    CHECK(larryArray({3, 2, 1, 5, 6}) == false);
    CHECK(larryArray({1, 2, 3, 5, 4}) == false);        
    CHECK(larryArray({3, 1, 2}) == true);
    CHECK(larryArray({1, 3, 4, 2}) == true);

}

File: Month-2/Week-6/day-40/Go/main.go
package main

import (
	"fmt"
)

func main() {

	input := []int{1, 6, 5, 2, 4, 3}
	fmt.Println(rotateArray(input))
}

func rotateArray(larryArray []int) (result bool) {
	inversions := 0

	for i := 0; i < len(larryArray); i++ {
		for j := i + 1; j < len(larryArray); j++ {
			if larryArray[i] > larryArray[j] {
				inversions++
			}
		}
	}

	if inversions%2 == 0 {
		result = true
	} else {
		result = false
	}
	return
}
